File: cpp/tests/temperature/tmp117_test_linux/tmp117_test_linux.cpp
#ifndef TEST_I2C_BUS
#define TEST_I2C_BUS 1
#endif

#include <stdio.h>
#include <unistd.h>
#include "I2CConnectionLinux.h"
#include "TMP117.h"

static int passed = 0, failed = 0;

static void check_true(bool cond, const char *label) {
    if (cond) { printf("PASS %s\n", label); passed++; }
    else       { printf("FAIL %s\n", label); failed++; }
}

int main() {
    I2CConnectionLinux connection(TEST_I2C_BUS, TMP117Minimal::I2C_ADDRESS);

    TMP117Minimal sensor(connection);                       // Create TMP117 driver, (connection)
    TMP117Full full(connection);                            // Create TMP117 Full driver, (connection)
    full.configure(TMP117Full::Mode::Continuous, 8, 0.125f);  // Configure conversion, (mode=Continuous, averaging=8, cycleSeconds=1.0 s) → bool
    usleep(300UL * 1000UL);
    float t = sensor.readTemperature();                     // Read temperature, () → °C
    check_true(t >= -40.0f && t <= 125.0f, "temperature_plausible");

    full.configure(TMP117Full::Mode::Continuous, 32, 4.0f);  // Configure conversion, (mode=Continuous, averaging=8, cycleSeconds=1.0 s) → bool
    TMP117Full::Config cfg = full.getConfig();              // Read conversion config, () → Config
    check_true(cfg.mode == TMP117Full::Mode::Continuous && cfg.averaging == 32 && cfg.cycleSeconds == 4.0f,
               "config_roundtrip");
    full.configure(TMP117Full::Mode::Shutdown, 0, 0.0155f);  // Configure conversion, (mode=Continuous, averaging=8, cycleSeconds=1.0 s) → bool
    check_true(full.isShutdown(), "shutdown");
    full.triggerOneShot();                                  // Start one conversion, () → void
    usleep(50UL * 1000UL);
    check_true(full.isDataReady(), "one_shot_data_ready");
    check_true(full.isShutdown(), "one_shot_returns_to_shutdown");
    full.configure();                                       // Configure conversion, (mode=Continuous, averaging=8, cycleSeconds=1.0 s) → bool
    check_true(!full.isShutdown(), "continuous");

    full.setHighLimit(80.0f);                               // Set THIGH_LIMIT, (celsius °C) → void
    check_true(full.getHighLimit() == 80.0f, "high_limit_roundtrip");
    full.setLowLimit(-10.25f);                              // Set TLOW_LIMIT, (celsius °C) → void
    check_true(full.getLowLimit() == -10.25f, "low_limit_roundtrip");
    full.setTemperatureOffset(0.5f);                        // Set calibration offset, (celsius °C) → void
    check_true(full.getTemperatureOffset() == 0.5f, "offset_roundtrip");
    full.setTemperatureOffset(0.0f);                        // Set calibration offset, (celsius °C) → void

    check_true(!full.isEepromBusy(), "eeprom_not_busy");
    uint16_t eeprom2 = 0;
    full.readEepromScratch(2, eeprom2);                     // Read EEPROM scratch, (slot 1|2|3, value&) → bool
    full.writeEepromScratch(2, 0xA55A);                     // Write EEPROM scratch, (slot 2, value 16-bit) → bool
    uint16_t scratch = 0;
    check_true(full.readEepromScratch(2, scratch) && scratch == 0xA55A, "eeprom2_volatile_roundtrip");

    // One real EEPROM program cycle (the conformance-checked eeprom_write_ready
    // timing): rewrite EEPROM2's original value while unlocked, so the stored
    // power-on value is unchanged. Costs one EEPROM2 endurance cycle per run.
    full.unlockEeprom();                                    // Unlock EEPROM, () → void
    full.writeEepromScratch(2, eeprom2);                    // Write EEPROM scratch, (slot 2, value 16-bit) → bool
    int waitedMs = 0;
    while (full.isEepromBusy() && waitedMs < 50) {          // Check EEPROM busy, () → bool
        usleep(1000UL);
        waitedMs += 1;
    }
    full.lockEeprom();                                      // Lock EEPROM, () → void
    check_true(waitedMs < 50, "eeprom_write_ready");
    check_true(full.readEepromScratch(2, scratch) && scratch == eeprom2, "eeprom2_restored");

    // High limit below ambient forces HIGH_Alert on the next conversion; in
    // Alert mode the flag latches until CONFIGURATION is read.
    full.configureAlert();                                  // Configure ALERT, (mode=Alert, polarity=ActiveLow, pinFunction=Alert) → void
    full.configure(TMP117Full::Mode::Continuous, 0, 0.0155f);  // Configure conversion, (mode=Continuous, averaging=8, cycleSeconds=1.0 s) → bool
    full.setHighLimit(t - 20.0f);                           // Set THIGH_LIMIT, (celsius °C) → void
    usleep(100UL * 1000UL);
    check_true((full.pollInterrupt() & TMP117Full::SOURCE_HIGH) != 0, "poll_interrupt_high");
    full.setHighLimit(80.0f);                               // Set THIGH_LIMIT, (celsius °C) → void
    usleep(100UL * 1000UL);
    full.pollInterrupt();                                   // Read alert flags, () → uint8_t mask
    check_true((full.pollInterrupt() & TMP117Full::SOURCE_HIGH) == 0, "poll_interrupt_clear");

    // Soft reset reloads CONFIGURATION, the limits and the offset from EEPROM.
    full.reset();                                           // Software reset, () → void
    check_true(full.getConfig().mode == TMP117Full::Mode::Continuous, "reset_restores_config");

    printf("===DONE: %d passed, %d failed===\n", passed, failed);
    return failed == 0 ? 0 : 1;
}

File: python/tests/temperature/tmp117_test_linux.py
"""Linux hardware test for the TMP117 — runs against a Raspberry Pi.

Checks the identity, a plausible reading, conversion-config/limit/offset
round-trips, Shutdown and one-shot conversion, the volatile EEPROM2 scratch
register, one EEPROM program cycle (EEPROM2 rewritten with its own value),
the HIGH_Alert flag, and soft reset. No power-on default is changed.
"""

import os
import time
from periph.connection.i2c_linux import I2CConnection
from periph.chips.temperature.tmp117 import TMP117Minimal, TMP117Full, SOURCE_HIGH, I2C_ADDRESS

passed = 0
failed = 0


def check_true(label, condition):
    global passed, failed
    if condition:
        print('PASS', label)
        passed += 1
    else:
        print('FAIL', label)
        failed += 1

I2C_BUS = int(os.environ.get('LINUX_I2C_BUS', '1'))
I2C_ADDR = int(os.environ.get('I2C_ADDR', hex(I2C_ADDRESS)), 16)

connection = I2CConnection(I2C_BUS, I2C_ADDR)

sensor = TMP117Minimal(connection)
check_true('construct_minimal', isinstance(sensor, TMP117Minimal))
full = TMP117Full(connection)
full.configure(mode='continuous', averaging=8, cycle_seconds=0.125)
time.sleep(0.3)
t = sensor.read_temperature()
check_true('temperature_plausible', -40.0 <= t <= 125.0)

full.configure(mode='continuous', averaging=32, cycle_seconds=4.0)
check_true('config_roundtrip', full.get_config() == ('continuous', 32, 4.0))
full.configure(mode='shutdown', averaging=0, cycle_seconds=0.0155)
check_true('shutdown', full.is_shutdown())
full.trigger_one_shot()
time.sleep(0.05)
check_true('one_shot_data_ready', full.is_data_ready())
check_true('one_shot_returns_to_shutdown', full.is_shutdown())
full.configure()
check_true('continuous', not full.is_shutdown())

full.set_high_limit(80.0)
check_true('high_limit_roundtrip', full.get_high_limit() == 80.0)
full.set_low_limit(-10.25)
check_true('low_limit_roundtrip', full.get_low_limit() == -10.25)
full.set_temperature_offset(0.5)
check_true('offset_roundtrip', full.get_temperature_offset() == 0.5)
full.set_temperature_offset(0.0)

check_true('eeprom_not_busy', not full.is_eeprom_busy())
eeprom2 = full.read_eeprom_scratch(2)
full.write_eeprom_scratch(2, 0xA55A)
check_true('eeprom2_volatile_roundtrip', full.read_eeprom_scratch(2) == 0xA55A)

# One real EEPROM program cycle (the conformance-checked eeprom_write_ready
# timing): rewrite EEPROM2's original value while unlocked, so the stored
# power-on value is unchanged. Costs one EEPROM2 endurance cycle per run.
full.unlock_eeprom()
full.write_eeprom_scratch(2, eeprom2)
waited_ms = 0
while full.is_eeprom_busy() and waited_ms < 50:
    time.sleep(0.001)
    waited_ms += 1
full.lock_eeprom()
check_true('eeprom_write_ready', waited_ms < 50)
check_true('eeprom2_restored', full.read_eeprom_scratch(2) == eeprom2)

# High limit below ambient forces HIGH_Alert on the next conversion; in
# Alert mode the flag latches until CONFIGURATION is read.
full.configure_alert(mode='alert', polarity='active_low', pin_function='alert')
full.configure(mode='continuous', averaging=0, cycle_seconds=0.0155)
full.set_high_limit(t - 20.0)
time.sleep(0.1)
check_true('poll_interrupt_high', full.poll_interrupt() & SOURCE_HIGH != 0)
full.set_high_limit(80.0)
time.sleep(0.1)
full.poll_interrupt()
check_true('poll_interrupt_clear', full.poll_interrupt() & SOURCE_HIGH == 0)

# Soft reset reloads CONFIGURATION, the limits and the offset from EEPROM.
full.reset()
check_true('reset_restores_config', full.get_config()[0] == 'continuous')

print('===DONE: %d passed, %d failed===' % (passed, failed))
